File: MarketSystem.Domain/Entities/RegistrationRequest.cs
using MarketSystem.Domain.Common;
using MarketSystem.Domain.Enums;

namespace MarketSystem.Domain.Entities;

/// <summary>
/// Public sign-up request. An anonymous visitor submits FullName + Phone; a
/// SuperAdmin later approves or rejects from the admin console. On approval a
/// new <see cref="User"/> (Role=Owner) and a dedicated <see cref="Market"/>
/// are created so that the owner's data stays isolated from every other tenant.
/// </summary>
public class RegistrationRequest : BaseEntity
{
    public string FullName { get; set; } = string.Empty;
    public string Phone { get; set; } = string.Empty;

    public RegistrationRequestStatus Status { get; set; } = RegistrationRequestStatus.Pending;
    // CreatedAt inherited from BaseEntity.
    public DateTime? ProcessedAt { get; set; }

    /// <summary>
    /// Optimistic-concurrency token (PostgreSQL system column xmin). Two SuperAdmins
    /// trying to approve the same request simultaneously will collide here and one
    /// of them will get a DbUpdateConcurrencyException → 409 instead of silently
    /// producing duplicate User/Market rows.
    /// </summary>
    public uint Xmin { get; set; }

    /// <summary>The SuperAdmin who approved or rejected this request.</summary>
    public Guid? ProcessedByUserId { get; set; }
    public User? ProcessedByUser { get; set; }

    /// <summary>The owner user created when this request was approved.</summary>
    public Guid? CreatedUserId { get; set; }
    public User? CreatedUser { get; set; }

    /// <summary>The market created when this request was approved.</summary>
    public int? CreatedMarketId { get; set; }
    public Market? CreatedMarket { get; set; }

    /// <summary>Optional rejection reason (shown back to the SuperAdmin only).</summary>
    public string? RejectReason { get; set; }
}

File: MarketSystem.Client/lib/features/superadmin/domain/models/registration_request.dart
/// Mirror of the backend's `RegistrationRequestDto` (see
/// `MarketSystem.Application/DTOs/RegistrationRequestDTOs.cs`). Plain
/// data class — no business logic.
class RegistrationRequest {
  RegistrationRequest({
    required this.id,
    required this.fullName,
    required this.phone,
    required this.status,
    required this.createdAt,
    this.processedAt,
    this.processedByUserName,
    this.createdUserId,
    this.createdMarketId,
    this.rejectReason,
  });

  final String id;
  final String fullName;
  final String phone;
  final String status; // "Pending" | "Approved" | "Rejected"
  final DateTime createdAt;
  final DateTime? processedAt;
  final String? processedByUserName;
  final String? createdUserId;
  final int? createdMarketId;
  final String? rejectReason;

  bool get isPending => status == 'Pending';
  bool get isApproved => status == 'Approved';
  bool get isRejected => status == 'Rejected';

  factory RegistrationRequest.fromJson(Map<String, dynamic> json) {
    return RegistrationRequest(
      id: json['id'] as String,
      fullName: json['fullName'] as String? ?? '',
      phone: json['phone'] as String? ?? '',
      status: json['status'] as String? ?? 'Pending',
      createdAt: DateTime.parse(json['createdAt'] as String).toUtc(),
      processedAt: json['processedAt'] != null
          ? DateTime.parse(json['processedAt'] as String).toUtc()
          : null,
      processedByUserName: json['processedByUserName'] as String?,
      createdUserId: json['createdUserId'] as String?,
      createdMarketId: json['createdMarketId'] as int?,
      rejectReason: json['rejectReason'] as String?,
    );
  }
}
